lib/bandit/http1/adapter.ex

```diff
@@ -194,87 +194,42 @@ defmodule Bandit.HTTP1.Adapter do
     {:ok, <<>>, %{req | metrics: metrics}}
   end
 
-  def read_req_body(
-        %__MODULE__{state: :headers_read, buffer: buffer, body_remaining: 0} = req,
-        _opts
-      ) do
-    time = Bandit.Telemetry.monotonic_time()
-
-    metrics =
-      req.metrics
-      |> Map.update(:req_body_bytes, byte_size(buffer), &(&1 + byte_size(buffer)))
-      |> Map.put_new(:req_body_start_time, time)
-      |> Map.put(:req_body_end_time, time)
-
-    {:ok, buffer, %{req | state: :body_read, buffer: <<>>, metrics: metrics}}
-  end
-
-  def read_req_body(
-        %__MODULE__{state: :headers_read, body_remaining: body_remaining, buffer: buffer} = req,
-        opts
-      )
+  def read_req_body(%__MODULE__{state: :headers_read, body_remaining: body_remaining} = req, opts)
       when is_number(body_remaining) do
-    max_desired_bytes = Keyword.get(opts, :length, 8_000_000)
-    read_size = Keyword.get(opts, :read_length, 1_000_000)
-    read_timeout = Keyword.get(opts, :read_timeout)
-
-    if byte_size(buffer) >= max_desired_bytes do
-      <<to_return::binary-size(max_desired_bytes), rest::binary>> = buffer
-
-      metrics =
-        req.metrics
-        |> Map.update(:req_body_bytes, byte_size(to_return), &(&1 + byte_size(to_return)))
-        |> Map.put_new_lazy(:req_body_start_time, &Bandit.Telemetry.monotonic_time/0)
-
-      {:more, to_return, %{req | buffer: rest, metrics: metrics}}
-    else
-      metrics =
-        Map.put_new_lazy(req.metrics, :req_body_start_time, &Bandit.Telemetry.monotonic_time/0)
-
-      to_read = min(body_remaining, max_desired_bytes - byte_size(buffer))
-
-      with {:ok, iolist} <- read(req.socket, to_read, [], read_size, read_timeout) do
-        result = IO.iodata_to_binary([buffer | iolist])
-        result_size = byte_size(result)
-        body_remaining = body_remaining - result_size
+    metrics =
+      Map.put_new_lazy(req.metrics, :req_body_start_time, &Bandit.Telemetry.monotonic_time/0)
 
-        if body_remaining > 0 do
-          metrics = Map.update(metrics, :req_body_bytes, result_size, &(&1 + result_size))
+    with {:ok, to_return, buffer, body_remaining} <-
+           do_read_content_length_body(req.socket, req.buffer, body_remaining, opts) do
+      if buffer == <<>> && body_remaining == 0 do
```

Comment:
I'm not so sure about comparing with empty binary as efficient as calling to `byte_size/1`'s BIF

Reply:
Yep, good call.